### ISSUE TITLE
ConfigInjectionBean should not be an alternative bean

### ISSUE THREAD

**Issue by mkouba** [wildfly-extras/wildfly-microprofile-config#26]
* https://github.com/wildfly-extras/wildfly-microprofile-config/blame/master/implementation/src/main/java/org/wildfly/microprofile/config/inject/ConfigInjectionBean.java#L167

As a result, any `ConfigInjectionBean` added in `org.wildfly.microprofile.config.inject.ConfigExtension.registerConfigProducer()` is ignored unless it is explicitly selected for a bean archive.

The original author of the bean was probably testing the code with some old version of OWB - see also https://issues.apache.org/jira/browse/OWB-1209.

**Comment by mkouba** [comment]
Note that MP config TCK [ConverterTest](https://github.com/eclipse/microprofile-config/blob/master/tck/src/main/java/org/eclipse/microprofile/config/tck/ConverterTest.java) fails for WildFly Swarm because of this issue.

**Comment by jmesnil** [comment]
@mkouba thanks for the report, I did the change in this branch https://github.com/jmesnil/wildfly-microprofile-config/tree/ConfigInjectionBean and also took the opportunity to use Swarm as the Arquillian container instead of OWB (so that the TCK is run with Weld).

However, I have now errors with that setting:

```
2017-11-29 10:17:55,856 INFO  [org.jboss.weld.Bootstrap] (Weld Thread Pool -- 4) WELD-000119: Not generating any bean definitions from org.wildfly.microprofile.config.SysPropConfigSource because of underlying class loading error: Type Failed to link org.wildfly.microprofile.config.SysPropConfigSource (Module "deployment.cdiOptionalInjectionTest.war:main" from Service Module Loader): org.eclipse.microprofile.config.spi.ConfigSource not found.  If this is unexpected, enable DEBUG logging to see the full error.
2017-11-29 10:17:55,856 INFO  [org.jboss.weld.Bootstrap] (Weld Thread Pool -- 9) WELD-000119: Not generating any bean definitions from org.wildfly.microprofile.config.WildFlyConfigBuilder because of underlying class loading error: Type Failed to link org.wildfly.microprofile.config.WildFlyConfigBuilder (Module "deployment.cdiOptionalInjectionTest.war:main" from Service Module Loader): org.eclipse.microprofile.config.spi.ConfigBuilder not found.  If this is unexpected, enable DEBUG logging to see the full error.
2017-11-29 10:17:55,856 INFO  [org.jboss.weld.Bootstrap] (Weld Thread Pool -- 3) WELD-000119: Not generating any bean definitions from org.wildfly.microprofile.config.EnvConfigSource because of underlying class loading error: Type Failed to link org.wildfly.microprofile.config.EnvConfigSource (Module "deployment.cdiOptionalInjectionTest.war:main" from Service Module Loader): org.eclipse.microprofile.config.spi.ConfigSource not found.  If this is unexpected, enable DEBUG logging to see the full error.
2017-11-29 10:17:55,856 INFO  [org.jboss.weld.Bootstrap] (Weld Thread Pool -- 6) WELD-000119: Not generating any bean definitions from org.wildfly.microprofile.config.DirConfigSource because of underlying class loading error: Type Failed to link org.wildfly.microprofile.config.DirConfigSource (Module "deployment.cdiOptionalInjectionTest.war:main" from Service Module Loader): org.eclipse.microprofile.config.spi.ConfigSource not found.  If this is unexpected, enable DEBUG logging to see the full error.
2017-11-29 10:17:55,857 INFO  [org.jboss.weld.Bootstrap] (Weld Thread Pool -- 8) WELD-000119: Not generating any bean definitions from org.wildfly.microprofile.config.PropertiesConfigSourceProvider because of underlying class loading error: Type Failed to link org.wildfly.microprofile.config.PropertiesConfigSourceProvider (Module "deployment.cdiOptionalInjectionTest.war:main" from Service Module Loader): org.eclipse.microprofile.config.spi.ConfigSourceProvider not found.  If this is unexpected, enable DEBUG logging to see the full error.
2017-11-29 10:17:55,857 INFO  [org.jboss.weld.Bootstrap] (Weld Thread Pool -- 5) WELD-000119: Not generating any bean definitions from org.wildfly.microprofile.config.inject.ConfigInjectionBean$ConfigPropertyLiteral because of underlying class loading error: Type Failed to link org.wildfly.microprofile.config.inject.ConfigInjectionBean$ConfigPropertyLiteral (Module "deployment.cdiOptionalInjectionTest.war:main" from Service Module Loader): org.eclipse.microprofile.config.inject.ConfigProperty not found.  If this is unexpected, enable DEBUG logging to see the full error.
2017-11-29 10:17:55,857 INFO  [org.jboss.weld.Bootstrap] (Weld Thread Pool -- 1) WELD-000119: Not generating any bean definitions from org.wildfly.microprofile.config.inject.ConfigProducer because of underlying class loading error: Type org.eclipse.microprofile.config.Config from [Module "deployment.cdiOptionalInjectionTest.war:main" from Service Module Loader] not found.  If this is unexpected, enable DEBUG logging to see the full error.
2017-11-29 10:17:55,858 INFO  [org.jboss.weld.Bootstrap] (Weld Thread Pool -- 6) WELD-000119: Not generating any bean definitions from org.wildfly.microprofile.config.WildFlyConfig because of underlying class loading error: Type Failed to link org.wildfly.microprofile.config.WildFlyConfig (Module "deployment.cdiOptionalInjectionTest.war:main" from Service Module Loader): org.eclipse.microprofile.config.Config not found.  If this is unexpected, enable DEBUG logging to see the full error.
2017-11-29 10:17:55,858 INFO  [org.jboss.weld.Bootstrap] (Weld Thread Pool -- 7) WELD-000119: Not generating any bean definitions from org.wildfly.microprofile.config.PropertiesConfigSource because of underlying class loading error: Type Failed to link org.wildfly.microprofile.config.PropertiesConfigSource (Module "deployment.cdiOptionalInjectionTest.war:main" from Service Module Loader): org.eclipse.microprofile.config.spi.ConfigSource not found.  If this is unexpected, enable DEBUG logging to see the full error.
2017-11-29 10:17:55,858 INFO  [org.jboss.weld.Bootstrap] (Weld Thread Pool -- 3) WELD-000119: Not generating any bean definitions from org.wildfly.microprofile.config.WildFlyConfigProviderResolver because of underlying class loading error: Type Failed to link org.wildfly.microprofile.config.WildFlyConfigProviderResolver (Module "deployment.cdiOptionalInjectionTest.war:main" from Service Module Loader): org.eclipse.microprofile.config.spi.ConfigProviderResolver not found.  If this is unexpected, enable DEBUG logging to see the full error.
2017-11-29 10:17:55,858 INFO  [org.jboss.weld.Bootstrap] (Weld Thread Pool -- 9) WELD-000119: Not generating any bean definitions from org.wildfly.microprofile.config.Converters because of underlying class loading error: Type org.eclipse.microprofile.config.spi.Converter from [Module "deployment.cdiOptionalInjectionTest.war:main" from Service Module Loader] not found.  If this is unexpected, enable DEBUG logging to see the full error.
2017-11-29 10:17:56,333 ERROR [org.jboss.msc.service.fail] (MSC service thread 1-2) MSC000001: Failed to start service jboss.deployment.unit."cdiOptionalInjectionTest.war".WeldStartService: org.jboss.msc.service.StartException in service jboss.deployment.unit."cdiOptionalInjectionTest.war".WeldStartService: Failed to start service
        at org.jboss.msc.service.ServiceControllerImpl$StartTask.run(ServiceControllerImpl.java:1904)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: org.jboss.weld.exceptions.DeploymentException: WELD-001408: Unsatisfied dependencies for type Optional<String> with qualifiers @Default
  at injection point [BackedAnnotatedParameter] Parameter 1 of [BackedAnnotatedMethod] @Inject public org.eclipse.microprofile.config.tck.OptionalValuesBean.setStringValue(Optional<String>)
  at org.eclipse.microprofile.config.tck.OptionalValuesBean.setStringValue(OptionalValuesBean.java:47)
WELD-001475: The following beans match by type, but none have matching qualifiers:
  - Producer Method [Optional<T>] with qualifiers [@ConfigurationValue @Any] declared as [[UnbackedAnnotatedMethod] @ConfigurationValue @Dependent @Produces org.wildfly.swarm.container.runtime.cdi.ConfigurationValueProducer.produceOptionalConfigValue(InjectionPoint)]

        at org.jboss.weld.bootstrap.Validator.validateInjectionPointForDeploymentProblems(Validator.java:359)
        at org.jboss.weld.bootstrap.Validator.validateInjectionPoint(Validator.java:281)
        at org.jboss.weld.bootstrap.Validator.validateGeneralBean(Validator.java:134)
        at org.jboss.weld.bootstrap.Validator.validateRIBean(Validator.java:155)
        at org.jboss.weld.bootstrap.Validator.validateBean(Validator.java:518)
        at org.jboss.weld.bootstrap.ConcurrentValidator$1.doWork(ConcurrentValidator.java:68)
        at org.jboss.weld.bootstrap.ConcurrentValidator$1.doWork(ConcurrentValidator.java:66)
        at org.jboss.weld.executor.IterativeWorkerTaskFactory$1.call(IterativeWorkerTaskFactory.java:63)
        at org.jboss.weld.executor.IterativeWorkerTaskFactory$1.call(IterativeWorkerTaskFactory.java:56)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
        at org.jboss.threads.JBossThread.run(JBossThread.java:320)
```

Do you have any idea why Weld is complaining (while it was working with OWB)?

**Comment by mkouba** [comment]
Hi Jeff, maybe we should use `arquillian-weld-embedded` to run tck in this repo. We're working on a separate configuration for Swarm (see also https://issues.jboss.org/browse/SWARM-1644). There were some more problems when running the TCK in container mode (see for example https://github.com/mkouba/wildfly-swarm/blob/SWARM-1644/testsuite/microprofile-tcks/config/src/test/java/org/wildfly/swarm/microprofile/config/tck/ConfigAuxiliaryArchiveAppender.java), but `CdiOptionalInjectionTest` does pass for me.

**Comment by jmesnil** [comment]
Thanks, that's a good idea to use `arquillian-weld-embedded` to run the TCK without having to depend on Swarm.

I updated my branch to do that in https://github.com/jmesnil/wildfly-microprofile-config/tree/ConfigInjectionBean  but I get new errors:

```
arquillianBeforeClass(org.eclipse.microprofile.config.tck.CustomConfigSourceTest)  Time elapsed: 1.227 sec  <<< FAILURE!
org.jboss.weld.exceptions.DeploymentException:
WELD-001409: Ambiguous dependencies for type Config with qualifiers @Default
  at injection point [BackedAnnotatedField] @Inject private org.eclipse.microprofile.config.tck.CustomConfigSourceTest.config
  at org.eclipse.microprofile.config.tck.CustomConfigSourceTest.config(CustomConfigSourceTest.java:0)
  Possible dependencies:
  - Producer Method [Config] with qualifiers [@Any @Default] declared as [[UnbackedAnnotatedMethod] @Produces org.wildfly.microprofile.config.inject.ConfigProdu
cer.getConfig(InjectionPoint)],
  - Producer Method [Config] with qualifiers [@Any @Default] declared as [[BackedAnnotatedMethod] @Produces org.wildfly.microprofile.config.inject.ConfigProduce
r.getConfig(InjectionPoint)]
```

I did some search but I did not find out why the same `getConfig(InjectionPoint)`method if declared as both `[BackedAnnotatedMethod]`  and `[UnbackedAnnotatedMethod]`

**Comment by mkouba** [comment]
`BackedAnnotatedMethod` means that Weld discovered a class which declares the producer in a bean archive. `UnbackedAnnotatedMethod` means that there is an extension which added a "synthetic" `AnnotatedType` during discovery.

This line is probably the cause:
https://github.com/jmesnil/wildfly-microprofile-config/blob/ConfigInjectionBean/tck/src/test/java/org/wildfly/microprofile/config/WildFlyConfigArchiveProcessor.java#L47

In your case, `ConfigExtension.beforeBeanDiscovery(BeforeBeanDiscovery, BeanManager)` adds `ConfigProducer` as a synthetic `AnnotatedType` and the `ConfigProducer` class is also discovered.

Since `wildfly-microprofile-config-implementation` is not a bean archive I would remove this line.

**Comment by jmesnil** [comment]
This is fixed by #27
